src/main/java/com/google/devtools/build/lib/skyframe/ToolchainResolutionFunction.java

```diff
@@ -428,44 +432,40 @@ private static void determineToolchainImplementations(
     // Determine the potential set of toolchains.
     Table<ConfiguredTargetKey, ToolchainTypeInfo, Label> resolvedToolchains =
         HashBasedTable.create();
-    ImmutableSet.Builder<ToolchainTypeInfo> requiredToolchainTypesBuilder = ImmutableSet.builder();
-    List<Label> missingToolchains = new ArrayList<>();
+    List<Label> missingMandatoryToolchains = new ArrayList<>();
     for (SingleToolchainResolutionKey key : registeredToolchainKeys) {
       SingleToolchainResolutionValue singleToolchainResolutionValue =
           (SingleToolchainResolutionValue)
               results.getOrThrow(key, InvalidToolchainLabelException.class);
-        if (singleToolchainResolutionValue == null) {
-          valuesMissing = true;
-          continue;
-        }
+      if (singleToolchainResolutionValue == null) {
+        valuesMissing = true;
+        continue;
+      }
 
-      if (singleToolchainResolutionValue.availableToolchainLabels().isEmpty()) {
-        // Save the missing type and continue looping to check for more.
-        // TODO(katre): Handle mandatory/optional.
-        missingToolchains.add(key.toolchainType().toolchainType());
-      } else {
+      if (!singleToolchainResolutionValue.availableToolchainLabels().isEmpty()) {
         ToolchainTypeInfo requiredToolchainType = singleToolchainResolutionValue.toolchainType();
-        requiredToolchainTypesBuilder.add(requiredToolchainType);
         resolvedToolchains.putAll(
             findPlatformsAndLabels(requiredToolchainType, singleToolchainResolutionValue));
+      } else if (key.toolchainType().mandatory()) {
+        // Save the missing type and continue looping to check for more.
+        missingMandatoryToolchains.add(key.toolchainType().toolchainType());
       }
+      // TODO(katre): track missing optional toolchains?
     }
 
-    if (!missingToolchains.isEmpty()) {
-      throw new UnresolvedToolchainsException(missingToolchains);
+    // Verify that all mandatory toolchain types have a toolchain.
+    if (!missingMandatoryToolchains.isEmpty()) {
+      throw new UnresolvedToolchainsException(missingMandatoryToolchains);
     }
 
     if (valuesMissing) {
       throw new ValueMissingException();
```

Comment:
Don't remember: does this just signify Skyframe restart?

Reply:
Yes, MissingValueException is caught at the top level, which then returns a null to signify a skyframe restart. It saves having to add in handling for returning null through all of the levels of this code.